Proposed v3.3.0 - Major Changes

codejp3 opened the pull request:
Added PHP compatibility versions after testing:
https://wordpress.org/support/topic/php-8-0-support-17/

Added conditional check for non-object scenarios:
https://wordpress.org/support/topic/error-when-no-404-php-defined-in-theme/

mujahidi commented on includes/functions.php:
```diff
@@ -81,6 +81,11 @@ function acft_enqueue_google_fonts_file() {
     global $post;
     
+    // anticipate 404 errors or non-object queries where there is no $post->ID, just return.
+    if ( is_404() || !is_object($post) ) {
```
I haven't tested the code yet but I think not enqueuing the Google fonts on 404 page could create issues for those users who have used ACF Typography in ACF Options. An example would be using the same header and footer as the rest of the site. We will likely need to do it differently. What are your thoughts?

codejp3 replied:
Yeah, I get that.

Don't merge this one. I've got it covered in a MAJOR pull request coming within the next day or two. It's everything I mentioned in this topic:
https://wordpress.org/support/topic/enqueue-fonts-styles-in-custom-function/

In order to accomplish those things, I ended up breaking your function for enqueue-ing Google fonts in half, using the first half as a new function, and it's called twice - to enqueue Google fonts, and by my new shortcode for inserting in-line font stylesheet links/styles code.

Part of that function separation involved code changes that should resolve this. It now follow this logic:

- Function now allows for an optional $post_id to be supplied.
- If no $post_id supplied, it'll try to grab it from **inside** the loop with global $post (what you originally had).
- If still no $post_id, it'll try to grab it from **outside** the loop with global $wp_query.
- If there's still no $post_id after all of that, it will only check for option fields, and return them as the "all_fields" variable without any array merging. 
- If a $post_id is present, it'll process post fields, block fields, option fields, and merge the arrays, and return the merged array as the "all_fields" variable (what you originally had).

Give me another day or two to finish up the edits. Not sure if it should be released as v3.3 or v4.0, but I'll leave that up to you.

codejp3 commented:
Here's the major overhaul, proposed version 3.3.0
I've done initial testing, but haven't thoroughly tested it yet.

Sorry it's a big one! 

Includes the fix discussed previously for getting ID of non-objects, and all the new features mentioned in this topic:
https://wordpress.org/support/topic/enqueue-fonts-styles-in-custom-function/

I opted to leave out converting URLs to directory paths (what I need for my PDF generation function), because it's such an odd and uncommon thing and really doesn't belong in the core plugin. Instead, I included a Filter Hook within the new shortcode, so that I can alter the output for my specific use outside of the acf-typography plugin.

Overall changes:
* [UPDATE] Fixed consistent textdomain usage for i18n translations.
* [UPDATE] Made sure displayed strings are all i18n translation-ready.
* [UPDATE] Added a few minor additional checks for some variables to prevent possible errors in the future.
* [BUG] Fixed 'Notice: Trying to get property ID of non-object' when enqueueing font files on non-objects.
* [NEW] Google Fonts can now be saved and served locally.
* [NEW] Admin Settings Option to choose between serving font stylesheets and font files Remotely or Locally.
* [NEW] Description and link provided for getting Google Fonts API key in Admin Settings page.
* [NEW] Shortcode to print in-line stylesheet link or style HTML code (acf_typography_stylesheet).
* [NEW] Filter hook for the new acf_typography_stylesheet shortcode to alter the output of the in-line stylesheet links/styles.
* [NEW] Tested up to PHP 8.2
* [UPDATE] Minor code cleanup for readability. 
* [UPDATE] Added new screenshot for new Admin Settings Page Options. 


I'll send you a message after you've had a chance to digest this one. While making these changes, I also noticed an opportunity to support auto-enqueueing stylesheets for User Options/Profiles, Menus/Menu Items, Widgets, Attachments, Taxonomies, and Comments.....basically everything ACF supports adding fields/field_groups to.  Depending on someone's use, some or all of these should be enqueueing stylesheets in wp-admin too. Want to talk to you directly about your interest in supporting these things, and the best way to go about it.

Cheers!

codejp3 commented:
One last change for this major update:

* [NEW] Plugin now 100% translation-ready and initial .pot file generated!

see: https://github.com/mujahidi/acf-typography/issues/24

Added necessary translators comments, declared textdomain, function to load textdomain translations, languages directory, and generated the initial .pot translation file. I also updated readme.txt/readme.md, and changelog.md to reflect these changes as well.

If you want any help understanding how to generate the .pot file every time there's a plugin update, let me know!

codejp3 commented:
Fixing a few issues after more testing.

* [BUG] Fixed 'Warning: failed to open stream: HTTP request failed' when bad API Key supplied, or other API connection issues. #27
* [NEW] Admin Settings checks and help messages for Google API Key/connection issues.

See: https://github.com/mujahidi/acf-typography/issues/27

codejp3 reviewed (commented):
Ready for Final Review.

Tested pretty extensively, but may need a few adjustments before being labeled "stable".

mujahidi commented:
@codejp3 Thank you so much for your hard work and contribution to this repository. I have a request. Can you please verify first before pushing your changes to this pull request? With so many commits, I think we ought to rebase them.

I still haven't been able to look at your change requests, and hopefully, I will do that in a couple of days.

It is possible I do not release all the changes in v3.3.0. We'll have to check and validate that nothing breaks thoroughly. I am looking to release multiple versions and patches. It is up for discussion though.

Thanks once again.

codejp3 commented:
@mujahidi What are your thoughts on the best way to handle these changes. There's more I'd like to add, but I want to get these current changes in the base core code before expanding even more.

We could break down each "type" of change into their own minor releases. Trickle the changes out one at a time.

I think that would work for the majority of changes, EXCEPT for the major change of saving/serving stylesheets/font files locally.

If this makes sense, I'll start a new branch of 3.2.4, and add a few small changes and push them your way. Then 3.2.5, 3.2.6 and so on. A little at a time. Let me know!
